### PR TITLE
Eager ignore

### PR DESCRIPTION
**Pull request by mixonic**
Paths may be ignored because they contain a large number of files. The `getDirs` method should not actually descend into those ignored directories. Specifically in Ember.js our `tmp/` and `node_modules/` directories may both be extremely large and have complex symlinking that trolls the `validatePaths` method.

Additionally this commit flattens the recursion in `getDirs`.

### CODE REVIEW

**Comment by mixonic** [comment]
Could use some guidance about what may be annoying travis. The error message is broken (https://github.com/yui/yuitest/pull/29) unfortunately. Test pass locally (or the relevant ones do, there are a number of intermittent errors that were present on master).

**Comment by okuryu** [comment]
My guess that yui/yuitest#29 doesn't resolves failing build of PR, because it seems to show different errors with patching yui/yuitest#29.
https://gist.github.com/okuryu/05aee6bfa5ff1b07b424

**Comment by mixonic** [comment]
@okuryu these failures seem very unrelated though?

**Comment by mixonic** [comment]
@okuryu looking at yuitest failures again, but they do not seem related here. I sense a disturbance in the npm versioning of a dependency.

This commit has a massive, huge impact on Ember and other codebases. Would appreciate any aid in getting it though, despite the slow-progress of these codebases.

**Comment by okuryu** [comment]
@mixonic Sorry for my delay! Please let me try to investigate this ASAP.

**Comment by dschmidt** [comment]
Any news? This change reduces our doc build times from about a minute to ~2secs, so I would really like to see this upstreamed.

**Comment by stefanpenner** [comment]
> Test pass locally (or the relevant ones do, there are a number of intermittent errors that were present on master).

A rebased version of this PR, consistently demonstrates this issue.

**Comment by stefanpenner** [comment]
@mixonic not quite sure yet why their is a failure, but it does appear your test case passes without your provided implementation.

**Comment by mixonic** [comment]
It seems `setUp` fails silently and doesn't stop the a suite from running :trollface: :trollface:

Some tests blindly run on the whole `input` dir, and since the test I added relies on a fixture that causes a failure those suites attempting to use all of `input/` fail.

**Comment by stefanpenner** [comment]
@mixonic you likely should create a unit test, targeting `Y.getDirs` directly.

something like:
```js
        var dirs = Y.getDirs(__dirname + '/input/with-symlink', {
          ignorePaths: [
            'a/b',
            'a/d',
            'c'
          ]
        });

            // paths: 'a',
            // ignorePaths: [
            //   'a/b',
            //   'a/d',
            //   'c'
            // ]
        Assert.isArray(dirs);
        console.log(dirs);
        Assert.areSame(1, dirs.length, 'Failed to retrieve all path options');
        Assert.areSame(dirs[0], 'hi');
```

**Comment by stefanpenner** [comment]
@mixonic once this is green (and reasonable) i'll make sure it gets merged.

File: tests/utils.js [hunk 1]
```diff
@@ -93,4 +118,32 @@ suite.add(new YUITest.TestCase({
...
+    'test: gets paths as array': function() {
+        process.chdir(__dirname);
+        var pathPrefix = __dirname + '/input/folders1';
+        var dirs = Y.getDirs(pathPrefix, []);
+        Assert.isArray(dirs);
+        Assert.areSame(2, dirs.length);
+        Assert.areNotSame(-1, dirs.indexOf(pathPrefix + '/one'), 'contains path /one');
+        Assert.areNotSame(-1, dirs.indexOf(pathPrefix + '/one/two'), 'contains path /one/two');
+    },
+    'test: ignores paths': function() {
```
Comment (mixonic) [inline tests/utils.js]: This test fails without the changes to `lib/utils.js`.

File: tests/utils.js [hunk 2]
```diff
@@ -74,6 +80,25 @@ suite.add(new YUITest.TestCase({
 
         Assert.isArray(options.paths, 'Failed to set path');
         Assert.areSame(3, options.paths.length, 'Failed to retrieve all path options');
+    },
+    'test: ignore paths': function () {
```
Comment (mixonic) [inline tests/utils.js]: This test asserts existing behavior for `ignorePaths`.

**Comment by mixonic** [comment]
@stefanpenner The changes here are intended to more efficiently build a list of directories by avoiding traversal of ignored paths. There should be no changed behavior.

However I've added a test for `getDirs` that shows it now looks for an array of ignored paths, and have confirmed that fails as expected (and the tests confirming old behavior pass) on master.

File: tests/parser.js [hunk 3]
```diff
@@ -15,9 +15,16 @@ var existsSync = fs.existsSync || path.existsSync;
 var suite = new YUITest.TestSuite({
     name: 'Parser Test Suite',
     setUp: function () {
+        process.chdir(__dirname);
         var json = (new Y.YUIDoc({
             quiet: true,
-            paths: ['input/'],
+            paths: [
```
Comment (mixonic) [inline tests/parser.js]: These tests were naively presuming all fixtures were valid. I've narrowed the list to ones required for passing tests.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -19,7 +19,11 @@ executable. No need to reinstall the app to test your changes.
 Running Unit Tests
 ------------------
 
-You are also able to run unit tests by executing `npm test`.
+The following dependencies are required to run tests:
+
+    $ npm install -g istanbul ytestrunner yuitest
+
+You are able to run unit tests by executing `npm test`.
```
Comment (okuryu) [inline CONTRIBUTING.md]: Why did you need to do this? I think `npm test` should works after run `npm install`.

Comment (mixonic) [inline CONTRIBUTING.md]: The `test` command presumes that `instanbul` is installed as a global executable:

https://github.com/yui/yuidoc/blob/master/package.json#L90

Comment (stefanpenner) [inline CONTRIBUTING.md]: `npm run *` prefixes `./node_modules/.bin/` on the path, which should include istanbul `https://github.com/yui/yuidoc/blob/master/package.json#L83`

Comment (mixonic) [inline CONTRIBUTING.md]: okie dokie, I'll attempt to remove it.

Comment (mixonic) [inline CONTRIBUTING.md]: donezo.

**Comment by okuryu** [comment]
Merged. I'll publish a new version to npm in this weekend.

**Comment by okuryu** [comment]
Published v0.10.0.
